barretenberg/cpp/src/barretenberg/commitment_schemes/shplonk/shplemini.hpp

```diff
@@ -196,13 +196,115 @@ template <typename Curve> class ShpleminiVerifier_ {
     using GeminiVerifier = GeminiVerifier_<Curve>;
 
   public:
+    struct ClaimBatch {
+        RefVector<Commitment> commitments;
+        RefVector<Fr> evaluations;
+        // scalar used for batching the claims, excluding the power of batching challenge \rho
+        Fr batch_scalar = 0;
+    };
+
+    /**
+     * @brief Logic to support batching opening claims for unshifted and shifted polynomials in Shplemini
+     * @details Stores references to the commitments/evaluations of unshifted and shifted polynomials to be batched
+     * opened via Shplemini. Aggregates the commitments and batching scalars for each batch into the corresponding
+     * containers for Shplemini. Computes the batched evaluation. Contains logic for computing the per-batch scalars
+     * used to batch each set of claims (see details below).
+     * @note This class performs the actual batching of the evaluations but not of the commitments. The latter are
+     * simply appended to a larger container, along with the scalars used to batch them. This is because Shplemini
+     * is optimized to perform a single batch mul that includes all commitments from each stage of the PCS. See
+     * description of ShpleminiVerifier for more details.
+     *
+     */
+    struct ClaimBatcher {
+        std::optional<ClaimBatch> unshifted; // commitments and evaluations of unshifted polynomials
```

Comment:
I prefer this `std::optional container` and `if (container)` to the  `if (container.size() > 0)` syntax I used in the PolynomialBatcher. If this stays as it is I'll plan to make that class match this one in that sense.

Reply:
I hated that too but after some thorough googling I'm convinced that this is a very idomatic usage of optionals. I think the problem with the sumcheck use case was (1) it was a bad reason (just bc the prover and verifier shared a struct I think?) and (2) the external user had to know about the presence of optionals. Here its all encapsulated in the class and no one else has to know